org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/generators/xml2dtd/XML2DTDGeneratorTest.java

```diff
@@ -0,0 +1,141 @@
+		String dtd = "<!ELEMENT note EMPTY>" + lineSeparator() + //
+				"<!ATTLIST note version CDATA #IMPLIED>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void empty() throws IOException {
+		String xml = "<note />";
+		String dtd = "<!ELEMENT note EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void onlyCharacterContent() throws IOException {
+		String xml = "<note>ABCD</note>";
+		String dtd = "<!ELEMENT note (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void onlyElementChildren() throws IOException {
+		String xml = "<note><from>ABCD</from></note>";
+		String dtd = "<!ELEMENT note (from)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void characterContentAndElementChildren() throws IOException {
+		String xml = "<note>ABCD<from>ABCD</from></note>";
+		String dtd = "<!ELEMENT note (#PCDATA|from)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void characterContentAndSeveralElementChildren() throws IOException {
+		String xml = "<note><from>ABCD</from><to</to></note>";
+		String dtd = "<!ELEMENT note (from, to)>" + lineSeparator() + //
+				"<!ELEMENT from (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT to EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void threeLevel() {
+		String xml = "<a>\r\n" + //
+				"	<b>\r\n" + //
+				"		<c/>\r\n" + //
+				"	</b>\r\n" + //
+				"</a>";
+		String dtd = "<!ELEMENT a (b)>" + lineSeparator() + //
+				"<!ELEMENT b (c)>" + lineSeparator() + //
+				"<!ELEMENT c EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+
+	@Test
+	public void cardinalityPlus() {
+		String xml = "<a>\r\n" + // 
+				"	<b>\r\n" + //
+				"		<c />\r\n" + // 
+				"		<d />\r\n" + //
+				"	</b>\r\n" + //
+				"	<b>\r\n" + //
+				"		<c />\r\n" + // 
+				"		<d />\r\n" + //
+				"	</b>\r\n" + //
+				"</a>";
+		String dtd = "<!ELEMENT a (b+)>" + lineSeparator() + //
+				"<!ELEMENT b (c, d)>" + lineSeparator() + //
+				"<!ELEMENT c EMPTY>" + lineSeparator() + //
+				"<!ELEMENT d EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);		
+	}
+	
+	@Test
+	public void complexLevel() {
+		String xml = "<a>\r\n" + //
+				"	<b>\r\n" + //
+				"		<b>\r\n" + //
+				"			<d>X</d>\r\n" + //
+				"		</b>\r\n" + //
+				"	</b>\r\n" + //
+				"	<b>\r\n" + //
+				"		<c />\r\n" + //
+				"	</b>\r\n" + //
+				"</a>";
+		String dtd = "<!ELEMENT a (b+)>" + lineSeparator() + //
+				"<!ELEMENT b (b?, d?, c?)>" + lineSeparator() + //
+				"<!ELEMENT d (#PCDATA)>" + lineSeparator() + //
+				"<!ELEMENT c EMPTY>";
+		assertGrammarGenerator(xml, new DTDGeneratorSettings(), dtd);
+	}
+

```

Comment:
I don't know the state of cardinality, but maybe something like:

```xml
<a>
  <b>
    <c />
    <c />
  </b>
  <b />
  <b>
    <c />
 </b>
```

If cardinality is being pushed back for now, then ignore this.

Reply:
I add a test for this usecase (see severalSameElement)